scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java

```diff
@@ -131,30 +133,63 @@ public void processElement(ProcessContext c, BoundedWindow window) {
     }
 
     final UUID uuid = UUID.randomUUID();
+
+    // element already has an in-flight request
+    F inFlight = inFlightRequests.get(input);
+    if (inFlight != null) {
+      try {
+        futures.computeIfAbsent(
+            uuid,
+            key -> addCallback(
+                inFlight,
+                output -> {
+                  results.add(new Result(input, success(output), key, c.timestamp(), window));
+                  return null;
+                },
+                throwable -> {
+                  results.add(new Result(input, failure(throwable), key, c.timestamp(), window));
+                  return null;
+                }
+            )
+        );
+      } catch (Exception e) {
+        LOG.error("Failed to process element", e);
+        throw e;
+      }
+      requestCount++;
+      return;
+    }
+
     try {
       semaphore.acquire();
       try {
         futures.computeIfAbsent(
           uuid,
-          key -> addCallback(
-            asyncLookup((C) client.get(instanceId), input),
-            output -> {
-              semaphore.release();
-              try {
-                cacheSupplier.put(instanceId, input, output);
-                results.add(new Result(input, success(output), key, c.timestamp(), window));
-              } catch (Exception e) {
-                LOG.error("Failed to cache result", e);
-                throw e;
-              }
-              return null;
-            },
-            throwable -> {
-              semaphore.release();
-              results.add(new Result(input, failure(throwable), key, c.timestamp(), window));
-              return null;
-            }
-          )
+          key -> {
+            final F future = asyncLookup((C) client.get(instanceId), input);
+            final boolean shouldRemove = future.equals(inFlightRequests.computeIfAbsent(input, k -> future));
```

Comment:
info: why do we need this check here? it seems like we need to always remove a future from inflight map

Reply:
Because we only insert if there isn't already a future present for `input`; this avoids a specific race case in which a future might be removed 'early' by some other future for the same `input`

Reply:
Yeah, that's the idea, but to avoid too much locking & keeping the implementation simpler, I'm allowing a race where multiple threads can issue requests simultaneously. 

Basically this check keeps the Future around until completion instead of having another earlier-completing request removing it "early" at the cost of a single comparison.

Reply:
makes sense... an alternative to the `inFlightRequests` being a `ConcurrentMap<A, F>` would be as a `ConcurrentMap<UUID, Long>` which just maps the Future `UUID` to the # of times you should call `context.output` on in the `flush` method 🤔 . then we could simply remove keys from `inFlightRequests` at the same time you remove them from `futures` and there shouldn't be a race condition (I think, lol). wdyt?

there could be an issue if the same input has a different `Window` in different requests though.

Reply:
Eh, I'm disinclined to add any additional complexity. This change uses the rest of the code unchanged which is a benefit IMO.